### PR TITLE
Update workload when job changes before admission

### LINKED ISSUE #1038: Update workload when job changes before admission
The current workload design is that if the configuration of workload and job is found to be inconsistent, the current workload will be deleted and re-created.

If there are some special scenarios where I need to change the job configuration, it will cause my job to be re-queued, which is not very reasonable from a product perspective, the expectation is that the workload will not be re-created but will change as the job changes.

### PR DESCRIPTION
**Pull request by yaroslava-serdiuk**
#### What type of PR is this?
/kind feature

#### Which issue(s) this PR fixes:
Fixes #1038

#### Does this PR introduce a user-facing change?
```release-note
Jobs preserve their position in the queue if the number of pods change before being admitted
```

### CODE REVIEW

File: pkg/controller/jobframework/reconciler.go [hunk 1]
```diff
@@ -509,6 +519,21 @@ func (r *JobReconciler) equivalentToWorkload(job GenericJob, object client.Objec
...
+	if err != nil {
+		return nil, fmt.Errorf("can't construct workload for update: %w", err)
+	}
+	wl.Spec = newWl.Spec
+	if err = r.client.Update(ctx, wl); err != nil {
+		return nil, fmt.Errorf("updating existed workload: %w", err)
+	}
+
+	r.record.Eventf(object, corev1.EventTypeNormal, "UpdatedWorkload",
+		"Updated not matching Workload for suspended job: %v", wl)
```
Comment (mimowo) [inline pkg/controller/jobframework/reconciler.go:533]: More of a question: are we sure the job is suspended at this point? For example, i'm thinking if this code is executed when the job is running, but a user manually modified the admitted workload.

**Review by mimowo — COMMENTED** [review]
Generally LGTM, just a question regarding the scenario when the job is running.

File: pkg/controller/jobframework/reconciler.go (hunk 1 above)
Comment (yaroslava-serdiuk) [reply pkg/controller/jobframework/reconciler.go:533]: This function is called if the job is suspended and the workload is not admitted. Or do you mean that that state of the job was changed during the Reconcile() execution? I don't think we can do anything about it.

Comment (mimowo) [reply pkg/controller/jobframework/reconciler.go:533]: > This function is called if the job is suspended and the workload is not admitted.

I looked in the code a bit, but could not see any `if` that guarantees the `ensureOneWorkload` function is called only for suspended jobs.  Actually the comment `// If there's no workload exists and job is unsuspended, we'll stop it immediately.` suggests under some circumstances the job might be unsuspended. Please double check. 

If there are some scenarios in which the job is unsuspeded, maybe we should just modify the message as `"Updated not matching Workload for job: %v`.

Comment (yaroslava-serdiuk) [reply pkg/controller/jobframework/reconciler.go:533]: Right, the job which is passed to ensureOneWorkload could be unsuspended and then we may suspend it (line 427),  however for updating we check the status of job earlier (line 421).

If the job is unsuspended and workload is modified we unsuspend the job and recreate the workload. My understanding of the issue is that we want to update workloads only for suspended job.

Comment (mimowo) [reply pkg/controller/jobframework/reconciler.go:533]: Oh, I see, you actually have this check inside the function `job.IsSuspended() && !workload.IsAdmitted(toDelete[0])` which makes sure the updating is only for suspended jobs.

**Review by alculquicondor — COMMENTED** [review]
I think we also need to modify the workload_webhook, as it currently has PodSets as immutable.

Also please add a integration test for Job that starts with one that doesn't fit in the quota, (record the creationTimestmap of the Worklad at this point) and gets admitted after being updated to a smaller size (ensure that the creationTimestamp didn't change)

**Comment by yaroslava-serdiuk** [comment]
@alculquicondor 
I added a test that modifies parallelism and verify the creation stamp doesn't changed. The pod template requests are immutable fields, so I wanted to write the test you suggested in a way that job doesn't tolerate spot taint, so the job can't fit and with the toleration job can run. However I hit that adding toleration to the job does't update workload (or recreate in the old/current logic): https://github.com/kubernetes-sigs/kueue/issues/1264

**Review by mimowo — COMMENTED** [review]
/lgtm

**Comment by alculquicondor** [comment]
/release-note-edit

```release-note
Jobs preserve their position in the queue if their resource requests change before being admitted
```

File: pkg/controller/jobs/job/job_controller_test.go [hunk 2]
```diff
@@ -874,6 +874,26 @@ func TestReconciler(t *testing.T) {
 			},
 			wantErr: jobframework.ErrNoMatchingWorkloads,
 		},
+		"non-matching non-admitted workload is updated": {
```
Comment (alculquicondor) [inline pkg/controller/jobs/job/job_controller_test.go:877]: Let's add a case for when the Workload has a quota reservation. In that case we would remove the existing Workload, right?

We might want to evict the Workload in the future, but that's a follow up.

File: pkg/controller/jobs/job/job_controller_test.go [hunk 3]
```diff
@@ -955,6 +975,39 @@ func TestReconciler(t *testing.T) {
...
+				*utiltesting.MakeWorkload("job", "ns").Finalizers(kueue.ResourceInUseFinalizerName).
+					PodSets(*utiltesting.MakePodSet(kueue.DefaultPodSetName, 10).Request(corev1.ResourceCPU, "1").Obj()).
+					Queue("test-queue-new").
+					Priority(0).
+					Labels(map[string]string{
+						controllerconsts.JobUIDLabel: "test-uid",
+					}).
+					Obj(),
+			},
+		},
```
Comment (alculquicondor) [inline pkg/controller/jobs/job/job_controller_test.go:1010]: add a test cases for changing the priority class

File: pkg/webhooks/workload_webhook.go [hunk 4]
```diff
@@ -344,9 +344,12 @@ func ValidateWorkloadUpdate(newObj, oldObj *kueue.Workload) field.ErrorList {
 	specPath := field.NewPath("spec")
 	statusPath := field.NewPath("status")
 	allErrs = append(allErrs, ValidateWorkload(newObj)...)
-	allErrs = append(allErrs, apivalidation.ValidateImmutableField(newObj.Spec.PodSets, oldObj.Spec.PodSets, specPath.Child("podSets"))...)
-	allErrs = append(allErrs, apivalidation.ValidateImmutableField(newObj.Spec.PriorityClassSource, oldObj.Spec.PriorityClassSource, specPath.Child("priorityClassSource"))...)
-	allErrs = append(allErrs, apivalidation.ValidateImmutableField(newObj.Spec.PriorityClassName, oldObj.Spec.PriorityClassName, specPath.Child("priorityClassName"))...)
+
+	if workload.HasQuotaReservation(oldObj) {
```
Comment (alculquicondor) [inline pkg/webhooks/workload_webhook.go:348]: What is the rationale for only checking `oldObj`?

File: pkg/webhooks/workload_webhook_test.go [hunk 5]
```diff
@@ -397,7 +397,7 @@ func TestValidateWorkloadUpdate(t *testing.T) {
 		wantErr       field.ErrorList
 	}{
 		"podSets should not be updated: count": {
```
Comment (alculquicondor) [inline pkg/webhooks/workload_webhook_test.go:399]:
```suggestion
		"podSets should not be updated when has quota reservation: count": {
```

**Comment by yaroslava-serdiuk** [comment]
> Jobs preserve their position in the queue if their resource requests change before being admitted

Resources requests are immutable, the user can't change them

**Comment by alculquicondor** [comment]
/release-note-edit
```release-note
Jobs preserve their position in the queue if the number of pods change before being admitted
```

File: test/integration/controller/jobs/job/job_controller_test.go [hunk 6]
```diff
@@ -1543,6 +1549,51 @@ var _ = ginkgo.Describe("Job controller interacting with scheduler", ginkgo.Orde
...
+				return createdJob.Spec.Suspend
+			}, util.Timeout, util.Interval).Should(gomega.Equal(ptr.To(true)))
+		})
+
+		wlLookupKey := types.NamespacedName{Name: workloadjob.GetWorkloadNameForJob(jobName), Namespace: ns.Name}
+		createdWorkload := util.AwaitAndVerifyCreatedWorkload(ctx, k8sClient, wlLookupKey, createdJob)
+		createdTime := createdWorkload.CreationTimestamp
+
+		newParallelism := int32(1)
+		createdJob.Spec.Parallelism = &newParallelism
```
Comment (alculquicondor) [inline test/integration/controller/jobs/job/job_controller_test.go:1578]:
```suggestion
		createdJob.Spec.Parallelism = ptr.To[int32](1)
```

File: pkg/webhooks/workload_webhook.go (hunk 4 above)
Comment (yaroslava-serdiuk) [reply pkg/webhooks/workload_webhook.go:348]: I think we want to update the workload that doesn't have a quota reservation and this is oldObj. 
I think if the update function is called for new object that has quota reservation and the old doesn't (not sure what the real case for that), it's should be the same as we call update first for new obj without quota reservation and then reserve the quota. 
My understanding why we check the quota is that we can't guarantee the quota for new object, so we want to do updates when the quota is not reserved. If the quota is fine for new object, than we are fine. 
Do I miss something?

File: pkg/controller/jobs/job/job_controller_test.go (hunk 2 above)
Comment (yaroslava-serdiuk) [reply pkg/controller/jobs/job/job_controller_test.go:877]: we have the case for that just above current test ("non-matching admitted workload is deleted")

File: pkg/webhooks/workload_webhook.go (hunk 4 above)
Comment (alculquicondor) [reply pkg/webhooks/workload_webhook.go:348]: Yeah, I think that's fine.
I just remembered that it is not possible to update Spec and Status at the same time.

**Review by alculquicondor — COMMENTED** [review]
/approve
/lgtm

File: pkg/controller/jobframework/reconciler.go [hunk 7]
```diff
@@ -505,6 +515,21 @@ func (r *JobReconciler) equivalentToWorkload(job GenericJob, object client.Objec
 	return true
 }
 
+func (r *JobReconciler) updateWorkloadToMatchJob(ctx context.Context, job GenericJob, object client.Object, wl *kueue.Workload) (*kueue.Workload, error) {
+	newWl, err := r.constructWorkload(ctx, job, object)
+	if err != nil {
+		return nil, fmt.Errorf("can't construct workload for update: %w", err)
+	}
+	wl.Spec = newWl.Spec
+	if err = r.client.Update(ctx, wl); err != nil {
```
Comment (kerthcet) [inline pkg/controller/jobframework/reconciler.go:524]: So the key point here is we'll replace the unadmitted workload with the newest spec. right?

Comment (yaroslava-serdiuk) [reply pkg/controller/jobframework/reconciler.go:524]: Correct. However note, that currently we can update only the amount of pods/containers: https://github.com/kubernetes-sigs/kueue/issues/1264

**Pull request merged**